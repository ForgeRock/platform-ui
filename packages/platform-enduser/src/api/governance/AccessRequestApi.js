/**
 * Copyright (c) 2023 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { generateIgaApi } from '@forgerock/platform-shared/src/api/BaseApi';
import encodeQueryString from '@forgerock/platform-shared/src/utils/encodeQueryString';

export function getUserRequests(userId, params, filter) {
  params.action = 'search';
  const queryString = encodeQueryString(params, true);
  return generateIgaApi().post(`/governance/user/${userId}/requests${queryString}`, { targetFilter: filter });
}

export function getUserApprovals(userId, params, filter) {
  params._action = 'search';
  const queryString = encodeQueryString(params, false);
  return generateIgaApi().post(`/governance/user/${userId}/approvals${queryString}`, { targetFilter: filter });
}

/**
 * Take action on a request
 * @param {String} requestId ID of request
 * @param {String} phaseName Phase name in which the request is
 * @param {String} action action being realized to the request
 * @param {String} comment Comment to leave on request
 * @param {Array} updatedActors Actor who the request will be forwarded to
 * @returns {Promise}
 */
export function requestAction(requestId, action, phaseName, comment, updatedActors) {
  const url = `/governance/requests/${requestId}?_action=${action}&phaseName=${phaseName}`;
  const requestBody = action === 'reassign' ? { comment, updatedActors } : { comment };
  return generateIgaApi().post(url, requestBody);
}

export function saveNewRequest(payload) {
  return generateIgaApi().post('governance/requests?_action=create', payload);
}

/**
 * Validate a request for invalid entries
 * @param {String} payload ID of request
 * @param {Array} payload containing request details to validate
 * @returns {Promise}
 */
export function validateRequest(payload) {
  const url = '/governance/requests?_action=validate';
  return generateIgaApi().post(url, payload);
}
