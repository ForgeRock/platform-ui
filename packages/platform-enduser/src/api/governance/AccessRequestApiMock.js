/**
 * Copyright (c) 2023 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

/* eslint-disable no-unused-vars */

import { cloneDeep, merge, isEmpty } from 'lodash';

const accessRequest = {
  id: '',
  requestType: '',
  request: {
    common: {
      priority: '',
      startDate: '2023-06-22T19:23:26+00:00',
      endDate: '2023-07-15T19:23:26+00:00',
    },
  },
  // This is who created the access request
  requester: {
    mail: '',
    givenName: '',
    sn: '',
    id: '',
    userName: '',
  },
  // This is who we are requesting access for
  user: {
    mail: '',
    givenName: '',
    sn: '',
    id: '',
    userName: '',
  },
  decision: {
    status: 'in-progress',
    outcome: null,
    completionDate: null,
    deadline: null,
    comments: [],
    phases: [
      {
        name: 'userApprove',
        type: 'request',
        status: 'in-progress',
        decision: null,
        startDate: '2023-06-15T18:14:56+00:00',
        events: {
          assignment: {
            notification: 'requestAssigned',
          },
          reassign: {
            notification: 'requestReassign',
          },
        },
        workflowTaskId: '2511',
      },
    ],
  },
};

const types = [
  {
    name: 'account',
    requestTypes: [
      'accountGrant',
      'accountRevoke',
    ],
  },
  {
    name: 'entitlement',
    requestTypes: [
      'entitlementGrant',
      'entitlementRevoke',
    ],
  },
  {
    name: 'role',
    requestTypes: [
      'roleGrant',
      'roleRevoke',
    ],
  },
];

const priorities = [
  'high',
  'medium',
  'low',
];

const statusMap = {
  completed: 'completed',
  pending: 'in-progress',
  cancelled: 'cancelled',
};

const users = [
  {
    givenName: 'Mike',
    id: '1234-456-1',
    mail: 'mike.wong@test.com',
    sn: 'Wong',
    userName: 'mike.wong@test.com',
  },
  {
    givenName: 'Andrew',
    id: '1234-456-2',
    mail: 'andrew.hertel@test.com',
    sn: 'Hertel',
    userName: 'andrew.hertel@test.com',
  },
  {
    givenName: 'Manuel',
    id: '1234-456-3',
    mail: 'manuel.escobar@test.com',
    sn: 'Escobar',
    userName: 'manuel.escobar@test.com',
  },
];

const applications = [
  {
    description: 'My Service Now App',
    icon: '',
    id: '1',
    name: 'My Service Now App',
    templateName: 'servicenow',
    templateVersion: '2.0',
  },
  {
    description: 'My Azure App',
    icon: '',
    id: '2',
    name: 'My Azure App',
    templateName: 'azure.ad',
    templateVersion: '2.0',
  },
];

const entitlements = [
  {
    displayName: 'All Users',
    description: 'This is for all users',
    id: '1',
  },
  {
    displayName: 'Groups Administrator',
    description: 'Administers different groups',
    id: '2',
  },
];

const roles = [
  {
    name: 'Main Role',
    description: 'The primary role for everyone',
    id: '1',
  },
  {
    name: 'Secondary Role',
    description: 'A secondary role with less permissions',
    id: '2',
  },
];

// const descriptors = [
//   {
//     idx: {
//       '/account': {
//         displayName: '',
//       },
//       '/entitlement': {
//         displayName: '',
//       },
//     },
//   },
// ];

function getDefaultRequest(values = {}) {
  const req = cloneDeep(accessRequest);
  merge(req, values);

  if (isEmpty(req.application)) delete req.application;
  if (isEmpty(req.entitlement)) delete req.entitlement;
  if (isEmpty(req.role)) delete req.role;

  return req;
}

function generateRequest(id, status) {
  const requesterInfo = users[id % users.length];
  const userInfo = users[(id + 1) % users.length];
  const type = types[id % types.length];
  const requestType = type.requestTypes[id % type.requestTypes.length];
  const request = {
    common: {
      priority: priorities[id % priorities.length],
    },
  };
  const decision = {
    status: statusMap[status],
  };

  let application = {};
  let entitlement = {};
  let role = {};

  switch (type.name) {
    case 'account':
      application = applications[id % applications.length];
      break;
    case 'entitlement':
      application = applications[id % applications.length];
      entitlement = entitlements[id % entitlements.length];
      break;
    case 'role':
      role = roles[id % roles.length];
      break;
    default:
      break;
  }

  return getDefaultRequest({
    id,
    requester: requesterInfo,
    user: userInfo,
    requestType,
    request,
    decision,
    application,
    entitlement,
    role,
  });
}

function generateRequests(pageSize, pageNumber, status) {
  if (pageNumber > 1) {
    return {
      totalCount: pageSize + 1,
      results: [
        generateRequest(pageSize + 1, status),
      ],
    };
  }

  const results = [];
  for (let i = 0; i < pageSize; i += 1) {
    results.push(generateRequest(i + 1, status));
  }

  return {
    totalCount: pageSize + 1,
    results,
  };
}

export function getRequestMock(id) {
  return generateRequest(id);
}

export function getRequestsMock(params, filters) {
  let pageSize = 10;
  let pageNumber = 1;
  let status = statusMap.pending;

  if (params.pageSize) pageSize = params.pageSize;
  if (params.pageNumber) pageNumber = params.pageNumber;
  if (params.status) status = params.status;

  return generateRequests(pageSize, pageNumber, status);
}
