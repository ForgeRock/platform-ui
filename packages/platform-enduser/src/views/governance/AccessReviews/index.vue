<!-- Copyright (c) 2023-2025 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <BContainer>
    <template v-if="tableLoading">
      <FrSpinner class="py-5" />
    </template>
    <div
      v-else
      class="my-5">
      <FrHeader
        :title="$t('pages.accessReview.title')"
        :subtitle="$t('pages.accessReview.subTitle')" />
      <BCard no-body>
        <BCardHeader class="p-0">
          <div class="btn-toolbar justify-content-between p-3 border-bottom-0">
            <div>
              <BDropdown
                id="dropdown-status"
                variant="link"
                class="ml-2 text-decoration-none">
                <template #button-content>
                  <span
                    class="font-weight-bold mr-1"
                  >
                    {{ `${$t('common.status')}:` }}
                  </span>
                  <span>{{ statusSort.text }}</span>
                </template>
                <template v-for="([key, statusObj]) in Array.from(statuses)">
                  <BDropdownItem
                    :key="key"
                    v-if="!statusObj.sortHide && !statusObj.sortAccessReviewHide"
                    @click="handleStatusSort(statusObj)"
                  >
                    <span
                      :class="[{'pl-3': (statusObj.type !== statuses.get('closed').type && statusObj.type !== statuses.get('in-progress').type)}]"
                    >
                      {{ statusObj.text }}
                    </span>
                  </BDropdownItem>
                </template>
              </BDropdown>
            </div>
          </div>
        </BCardHeader>
        <BTable
          v-if="accessReviewList.length"
          class="mb-0"
          hover
          responsive
          tbody-tr-class="cursor-pointer"
          primary-key="id"
          :fields="fields"
          :items="accessReviewList"
          @row-clicked="viewApplicationDetails">
          <template #cell(name)="{ item }">
            <BMedia
              class="align-items-center"
              no-body>
              <div class="media-body align-self-center">
                <h5 class="m-0">
                  {{ item.campaignName || item.name }}
                </h5>
                <small class="text-muted">
                  {{ item.description }}
                </small>
              </div>
            </BMedia>
          </template>
          <template #cell(status)="{ item }">
            <template v-if="item && item.status">
              <BBadge
                class="w-100"
                :variant="statuses.get(item.status).variant">
                {{ $t(`pages.accessReview.status.${item.status.toLowerCase()}`) }}
              </BBadge>
            </template>
          </template>
          <template #cell(progress)="{ item }">
            <FrCircleProgressBar
              :id="`taskProgress${item.id}`"
              :progress="progressPercentage(item.progress)"
              :thickness="6"
              :empty-thickness="6"
              :empty-color="styles.brightGray"
              :size="50">
              <template #count="{ count }">
                <small class="font-weight-bold mb-0">
                  {{ count.currentValue }}%
                </small>
              </template>
            </FrCircleProgressBar>
            <BPopover
              triggers="hover"
              boundary="window"
              placement="top"
              :target="`taskProgress${item.id}`">
              <div class="p-1">
                <h5 class="mb-1">
                  {{ $t('pages.accessReview.taskPopoverCompletedText', { percentage: progressPercentage(item.progress) }) }}
                </h5>
                {{ $t('pages.accessReview.taskPopoverItemsCompletedText', { completed: item.totals.completed, total: item.totals.total }) }}
              </div>
            </BPopover>
          </template>
        </BTable>
        <FrNoData
          v-else
          :card="false"
          class="mb-4"
          data-testid="access-review-no-data"
          icon="inbox"
          :subtitle="$t('governance.certificationTask.noAccessReview', { type: statusSort.text })" />
        <FrPagination
          v-if="totalRows"
          :value="currentPage"
          :total-rows="totalRows"
          :per-page="pageSize"
          @input="paginationChange"
          @on-page-size-change="pageSizeChange" />
      </BCard>
    </div>
  </BContainer>
</template>

<script>
import { mapState } from 'pinia';
import { useUserStore } from '@forgerock/platform-shared/src/stores/user';
import {
  BBadge,
  BCard,
  BCardHeader,
  BContainer,
  BDropdown,
  BDropdownItem,
  BMedia,
  BPopover,
  BTable,
} from 'bootstrap-vue';
import {
  getCertificationItems,
  searchCertificates,
  forwardCertification,
} from '@forgerock/platform-shared/src/api/governance/CertificationApi';
import NotificationMixin from '@forgerock/platform-shared/src/mixins/NotificationMixin';
import FrCircleProgressBar from '@forgerock/platform-shared/src/components/CircleProgressBar';
import FrHeader from '@forgerock/platform-shared/src/components/PageHeader';
import FrNoData from '@forgerock/platform-shared/src/components/NoData';
import FrPagination from '@forgerock/platform-shared/src/components/Pagination';
import FrSpinner from '@forgerock/platform-shared/src/components/Spinner';
import CertificationMixin from '@forgerock/platform-shared/src/mixins/Governance/Certification';
import styles from '@/scss/main.scss';

export default {
  name: 'AccessReviews',
  components: {
    BBadge,
    BCard,
    BCardHeader,
    BContainer,
    BDropdown,
    BDropdownItem,
    BMedia,
    BPopover,
    BTable,
    FrCircleProgressBar,
    FrHeader,
    FrNoData,
    FrPagination,
    FrSpinner,
  },
  data() {
    return {
      pageSize: 10,
      currentPage: 1,
      totalRows: 0,
      tableLoading: true,
      selectedCertId: null,
      accessReviewList: [],
      fields: [
        {
          key: 'name',
          label: this.$t('common.name'),
        },
        {
          key: 'formattedDeadline',
          label: this.$t('common.deadline'),
          sortable: true,
        },
        {
          key: 'status',
          class: 'w-140px',
          label: this.$t('common.status'),
        },
        {
          key: 'progress',
          class: 'w-140px',
          label: this.$t('common.progress'),
        }],
      styles,
    };
  },
  mixins: [
    CertificationMixin,
    NotificationMixin,
  ],
  mounted() {
    this.getList();
  },
  computed: {
    ...mapState(useUserStore, ['userId']),
    currentUserId() {
      return `managed/user/${this.userId}`;
    },
  },
  methods: {
    pageSizeChange(newPageSize) {
      this.pageSize = newPageSize;
      this.currentPage = 1;
      this.getList();
    },
    paginationChange(pageNumber) {
      this.currentPage = pageNumber;
      this.getList();
    },
    getItems(params) {
      const { queryString } = params;
      delete params.queryString;
      if (!queryString) {
        return getCertificationItems(params);
      }
      return searchCertificates(queryString, params);
    },
    viewApplicationDetails(rowData) {
      if (!rowData || !rowData.campaignId || !rowData.certifierId) {
        this.showErrorMessage('', this.$t('governance.certificationTask.errors.certificationDetailError'));
      } else {
        this.$router.push({
          name: 'CertificationTask',
          params: { campaignId: rowData?.campaignId },
          query: {
            actorId: rowData?.certifierId,
            taskStatus: this.statusSort.param,
          },
        });
      }
    },
    progressPercentage(progress = 0) {
      const actualProgress = progress * 100;
      return actualProgress >= 50
        ? Math.floor(actualProgress)
        : Math.ceil(actualProgress);
    },
    /**
     * @param {Object} data data of the certification to forward
     * @param {String} data.certId id of the certification to forward
     * @param {String} data.newActorId id of new actor
     * @param {String} data.comment comment for the forward
     */
    forward({ certId, newActorId, comment = '' }) {
      forwardCertification(this.currentUserId, certId, newActorId, comment)
        .then(() => {
          this.displayNotification('success', this.$t('governance.certificationTask.success.accessReviewForwarded'));
          this.$root.$emit('bv::hide::modal', 'forward-review-modal');
          this.getList();
        })
        .catch((error) => {
          this.showErrorMessage(error, this.$t('governance.certificationTask.errors.forwardItemError'));
        });
    },
  },
};
</script>
<style lang="scss" scoped>
#dropdown-status:deep {
  button {
    color: $black;
    padding-left: 0;
    padding-right: 0;
  }
}
:deep {
  .w-140px {
    width: 140px;
  }
 .w-96px {
    width: 96px;
  }
}
</style>
