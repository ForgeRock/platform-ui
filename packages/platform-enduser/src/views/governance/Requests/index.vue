<!-- Copyright (c) 2023 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <BContainer>
    <div class="mt-5">
      <FrHeader
        :title="$t('pageTitles.MyRequests')"
        :subtitle="$t('governance.accessRequests.myRequests.subtitle')" />
      <BCard
        no-body
        class="card-tabs-vertical" />
    </div>
  </BContainer>
</template>

<script>
import {
  BCard,
  BContainer,
} from 'bootstrap-vue';
import FrHeader from '@forgerock/platform-shared/src/components/PageHeader';
/**
 * @description Landing page for Requests
 */
export default {
  name: 'Requests',
  components: {
    BCard,
    BContainer,
    FrHeader,
  },
  data() {
    return {
    };
  },
};
</script>
