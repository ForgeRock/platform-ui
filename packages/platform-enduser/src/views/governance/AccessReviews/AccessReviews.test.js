/**
 * Copyright (c) 2023-2025 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { shallowMount, mount, flushPromises } from '@vue/test-utils';
import { findByTestId } from '@forgerock/platform-shared/src/utils/testHelpers';
import * as CertificationApi from '@forgerock/platform-shared/src/api/governance/CertificationApi';
import AccessReviews from './index';

describe('AccessReviews', () => {
  let wrapper;
  beforeEach(() => {
    CertificationApi.getCertificationItems = jest.fn().mockReturnValue(Promise.resolve({
      data: {
        test: 'test',
      },
    }));

    CertificationApi.searchCertificates = jest.fn().mockReturnValue(Promise.resolve({
      data: {
        test: 'test',
      },
    }));
  });

  it('should display noData component when no access reviews are found', async () => {
    wrapper = mount(AccessReviews, {
      global: {
        mocks: {
          $t: (id) => id,
          getItems: () => jest.fn().mockReturnValue(Promise.resolve({
            data: {
              test: 'test',
            },
          })),
        },
      },
    });
    wrapper.vm.setAccessReviewList({ result: [], totalCount: 0 });
    await flushPromises();

    const noData = findByTestId(wrapper, 'access-review-no-data');
    expect(noData.exists()).toBeTruthy();
  });

  it('Calling getList with not search params calls getCertificationItems', () => {
    wrapper = shallowMount(AccessReviews, {
      global: {
        mocks: {
          $t: (id) => id,
          getItems: () => jest.fn().mockReturnValue(Promise.resolve({
            data: {
              test: 'test',
            },
          })),
        },
      },
    });
    expect(CertificationApi.searchCertificates).not.toHaveBeenCalled();
    expect(CertificationApi.getCertificationItems).toHaveBeenCalledWith({
      pageNumber: 0, pageSize: 10, sortBy: 'deadline', sortDesc: false, status: 'active',
    });
  });

  it('Calling getList with not search params calls getCertificationItems', () => {
    wrapper = shallowMount(AccessReviews, {
      data() {
        return {
          searchQuery: 'test search',
        };
      },
      global: {
        mocks: {
          $t: (id) => id,
          getItems: () => jest.fn().mockReturnValue(Promise.resolve({
            data: {
              test: 'test',
            },
          })),
        },
      },
    });
    expect(CertificationApi.getCertificationItems).not.toHaveBeenCalled();
    expect(CertificationApi.searchCertificates).toHaveBeenCalledWith('test search', {
      pageNumber: 0, pageSize: 10, sortBy: 'deadline', sortDesc: false, status: 'active',
    });
  });
});
