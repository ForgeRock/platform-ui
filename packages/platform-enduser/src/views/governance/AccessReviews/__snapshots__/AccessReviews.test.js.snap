// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccessReviews Calling getList with not search params calls getCertificationItems: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "fluid" with v-bind value \`false\` will render fluid="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <BContainer>",
    "
",
    " at <AccessReviews",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;

exports[`AccessReviews should display noData component when no access reviews are found: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) Vue 3's render function API has changed. You can opt-in to the new API with:

  configureCompat({ RENDER_FUNCTION: false })

  (This can also be done per-component via the "compatConfig" option.)
  Details: https://v3-migration.vuejs.org/breaking-changes/render-function-api.html",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation GLOBAL_EXTEND) (2)",
    "
",
    " at <AccessReviews",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation COMPONENT_FUNCTIONAL) Functional component <BContainer> should be defined as a plain function in Vue 3. The "functional" option has been removed. NOTE: Before migrating to use plain functions for functional components, first make sure that all async components usage have been migrated and its compat behavior has been disabled.
  Details: https://v3-migration.vuejs.org/breaking-changes/functional-components.html",
    "
",
    " at <AccessReviews",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) vm.$listeners has been removed. In Vue 3, parent v-on listeners are included in vm.$attrs and it is no longer necessary to separately use v-on="$listeners" if you are already using v-bind="$attrs". (Note: the Vue 3 behavior only applies if this compat config is disabled)
  Details: https://v3-migration.vuejs.org/breaking-changes/listeners-removed.html",
    "
",
    " at <BContainer>",
    "
",
    " at <AccessReviews",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) (2)",
    "
",
    " at <BCard",
    "no-body=""",
    ">",
    "
",
    " at <BContainer>",
    "
",
    " at <AccessReviews",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) (3)",
    "
",
    " at <BCardHeader",
    "class="p-0"",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    ">",
    "
",
    " at <BContainer>",
    "
",
    " at <AccessReviews",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) (2)",
    "
",
    " at <BDropdown",
    "id="dropdown-status"",
    "variant="link"",
    "class="m-md-2 text-decoration-none"",
    ">",
    "
",
    " at <BCardHeader",
    "class="p-0"",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    ">",
    "
",
    " at <BContainer>",
    "
",
    " at <AccessReviews",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation OPTIONS_BEFORE_DESTROY) \`beforeDestroy\` has been renamed to \`beforeUnmount\`.",
    "
",
    " at <BDropdown",
    "id="dropdown-status"",
    "variant="link"",
    "class="m-md-2 text-decoration-none"",
    ">",
    "
",
    " at <BCardHeader",
    "class="p-0"",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    ">",
    "
",
    " at <BContainer>",
    "
",
    " at <AccessReviews",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_SCOPED_SLOTS) vm.$scopedSlots has been removed. Use vm.$slots instead.
  Details: https://v3-migration.vuejs.org/breaking-changes/slots-unification.html",
    "
",
    " at <BDropdown",
    "id="dropdown-status"",
    "variant="link"",
    "class="m-md-2 text-decoration-none"",
    ">",
    "
",
    " at <BCardHeader",
    "class="p-0"",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    ">",
    "
",
    " at <BContainer>",
    "
",
    " at <AccessReviews",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation GLOBAL_EXTEND) (3)",
    "
",
    " at <BDropdown",
    "id="dropdown-status"",
    "variant="link"",
    "class="m-md-2 text-decoration-none"",
    ">",
    "
",
    " at <BCardHeader",
    "class="p-0"",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    ">",
    "
",
    " at <BContainer>",
    "
",
    " at <AccessReviews",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation COMPONENT_FUNCTIONAL) (2)",
    "
",
    " at <BDropdown",
    "id="dropdown-status"",
    "variant="link"",
    "class="m-md-2 text-decoration-none"",
    ">",
    "
",
    " at <BCardHeader",
    "class="p-0"",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    ">",
    "
",
    " at <BContainer>",
    "
",
    " at <AccessReviews",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) (4)",
    "
",
    " at <BButton",
    "class="dropdown-toggle"",
    "onMousedown=fn<bound onMousedown>",
    "onClick=fn<bound toggle>",
    " ...",
    ">",
    "
",
    " at <BDropdown",
    "id="dropdown-status"",
    "variant="link"",
    "class="m-md-2 text-decoration-none"",
    ">",
    "
",
    " at <BCardHeader",
    "class="p-0"",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    ">",
    "
",
    " at <BContainer>",
    "
",
    " at <AccessReviews",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) (3)",
    "
",
    " at <BDropdownItem",
    "key="active"",
    "onClick=fn<onClick>",
    ">",
    "
",
    " at <BDropdown",
    "id="dropdown-status"",
    "variant="link"",
    "class="m-md-2 text-decoration-none"",
    ">",
    "
",
    " at <BCardHeader",
    "class="p-0"",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    ">",
    "
",
    " at <BContainer>",
    "
",
    " at <AccessReviews",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_SCOPED_SLOTS) (2)",
    "
",
    " at <BDropdownItem",
    "key="active"",
    "onClick=fn<onClick>",
    ">",
    "
",
    " at <BDropdown",
    "id="dropdown-status"",
    "variant="link"",
    "class="m-md-2 text-decoration-none"",
    ">",
    "
",
    " at <BCardHeader",
    "class="p-0"",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    ">",
    "
",
    " at <BContainer>",
    "
",
    " at <AccessReviews",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation GLOBAL_EXTEND) (4)",
    "
",
    " at <BDropdownItem",
    "key="active"",
    "onClick=fn<onClick>",
    ">",
    "
",
    " at <BDropdown",
    "id="dropdown-status"",
    "variant="link"",
    "class="m-md-2 text-decoration-none"",
    ">",
    "
",
    " at <BCardHeader",
    "class="p-0"",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    ">",
    "
",
    " at <BContainer>",
    "
",
    " at <AccessReviews",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) (4)",
    "
",
    " at <BLink",
    "class="dropdown-item"",
    "onClick=fn<bound onClick>",
    "ref="item"",
    " ...",
    ">",
    "
",
    " at <BDropdownItem",
    "key="active"",
    "onClick=fn<onClick>",
    ">",
    "
",
    " at <BDropdown",
    "id="dropdown-status"",
    "variant="link"",
    "class="m-md-2 text-decoration-none"",
    ">",
    "
",
    " at <BCardHeader",
    "class="p-0"",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    ">",
    "
",
    " at <BContainer>",
    "
",
    " at <AccessReviews",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation OPTIONS_DATA_MERGE) Detected conflicting key "bvListeners" when merging data option values. In Vue 3, data keys are merged shallowly and will override one another.
  Details: https://v3-migration.vuejs.org/breaking-changes/data-option.html#mixin-merge-behavior-change",
    "
",
    " at <BLink",
    "class="dropdown-item"",
    "onClick=fn<bound onClick>",
    "ref="item"",
    " ...",
    ">",
    "
",
    " at <BDropdownItem",
    "key="active"",
    "onClick=fn<onClick>",
    ">",
    "
",
    " at <BDropdown",
    "id="dropdown-status"",
    "variant="link"",
    "class="m-md-2 text-decoration-none"",
    ">",
    "
",
    " at <BCardHeader",
    "class="p-0"",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    ">",
    "
",
    " at <BContainer>",
    "
",
    " at <AccessReviews",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) (5)",
    "
",
    " at <BLink",
    "class="dropdown-item"",
    "onClick=fn<bound onClick>",
    "ref="item"",
    " ...",
    ">",
    "
",
    " at <BDropdownItem",
    "key="active"",
    "onClick=fn<onClick>",
    ">",
    "
",
    " at <BDropdown",
    "id="dropdown-status"",
    "variant="link"",
    "class="m-md-2 text-decoration-none"",
    ">",
    "
",
    " at <BCardHeader",
    "class="p-0"",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    ">",
    "
",
    " at <BContainer>",
    "
",
    " at <AccessReviews",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation OPTIONS_BEFORE_DESTROY) (2)",
    "
",
    " at <BLink",
    "class="dropdown-item"",
    "onClick=fn<bound onClick>",
    "ref="item"",
    " ...",
    ">",
    "
",
    " at <BDropdownItem",
    "key="active"",
    "onClick=fn<onClick>",
    ">",
    "
",
    " at <BDropdown",
    "id="dropdown-status"",
    "variant="link"",
    "class="m-md-2 text-decoration-none"",
    ">",
    "
",
    " at <BCardHeader",
    "class="p-0"",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    ">",
    "
",
    " at <BContainer>",
    "
",
    " at <AccessReviews",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_SCOPED_SLOTS) (3)",
    "
",
    " at <BLink",
    "class="dropdown-item"",
    "onClick=fn<bound onClick>",
    "ref="item"",
    " ...",
    ">",
    "
",
    " at <BDropdownItem",
    "key="active"",
    "onClick=fn<onClick>",
    ">",
    "
",
    " at <BDropdown",
    "id="dropdown-status"",
    "variant="link"",
    "class="m-md-2 text-decoration-none"",
    ">",
    "
",
    " at <BCardHeader",
    "class="p-0"",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    ">",
    "
",
    " at <BContainer>",
    "
",
    " at <AccessReviews",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_ATTRS_CLASS_STYLE) Component <BLink> has \`inheritAttrs: false\` but is relying on class/style fallthrough from parent. In Vue 3, class/style are now included in $attrs and will no longer fallthrough when inheritAttrs is false. If you are already using v-bind="$attrs" on component root it should render the same end result. If you are binding $attrs to a non-root element and expecting class/style to fallthrough on root, you will need to now manually bind them on root via :class="$attrs.class".
  Details: https://v3-migration.vuejs.org/breaking-changes/attrs-includes-class-style.html",
    "
",
    " at <BLink",
    "class="dropdown-item"",
    "onClick=fn<bound onClick>",
    "ref="item"",
    " ...",
    ">",
    "
",
    " at <BDropdownItem",
    "key="active"",
    "onClick=fn<onClick>",
    ">",
    "
",
    " at <BDropdown",
    "id="dropdown-status"",
    "variant="link"",
    "class="m-md-2 text-decoration-none"",
    ">",
    "
",
    " at <BCardHeader",
    "class="p-0"",
    ">",
    "
",
    " at <BCard",
    "no-body=""",
    ">",
    "
",
    " at <BContainer>",
    "
",
    " at <AccessReviews",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;
